New Branch with Completed NPC and World Item Classes

royjpark opened the pull request:
- Changed iNPC and iWorldItem's draw parameters to be the same as iLinkState.
- Added a Sprite Factory for both NPC and WorldItem.
- Created an Npc and Item class, comparable to Link class.
- Created classes for the following NPCs: skeleton, gel, bat, goriya (idk how to explain what it is lol), the boss, and the old man.
- Created classes for the following World Items: Triforce and Heart (the heart container I believe to be exact).

royjpark committed:
New Branch with completed npc and world item classes

sweeneyde commented on CrossPlatformDesktopProject/CrossPlatformDesktopProject/WorldItem/HeartIdle.cs:
```diff
@@ -0,0 +1,31 @@
+{
+    class HeartIdle : IWorldItem
+    {
+        private Item item;
```
Does this class actually use `item`? Or is this just for uniformity with other stuff?

bhasseman replied:
It doesn't look like `item` is used for anything, but in the future it may be that `item` has a specific type (in this case, it could be `heartItem`) which would affect its behavior (such as picking it up increases health)

royjpark replied:
Oh my bad, no it doesn't do anything so I removed it for now. But what Brent says makes sense.

sweeneyde commented on CrossPlatformDesktopProject/CrossPlatformDesktopProject/WorldItem/Item.cs:
```diff
@@ -0,0 +1,36 @@
+
+namespace CrossPlatformDesktopProject.WorldItem
+{
+    class Item
```
If there's only one "state", wouldn't it be simpler not to use the state pattern at all? I think the Link / Enemy stuff is different because there's one kind of `Player` object that can switch between different states, whereas a `Heart` can't change states to become a `Triforce` or vice-versa.

As in, we could have a `Heart` class that has a self-contained `Draw()` and empty `Update()` method, and then a `Triforce` class with the same interface? If the behavior needs to get more complicated in the future, that only needs to happen to those classes with multiple states.

Maybe I'm forgetting/misunderstanding something.

royjpark replied:
No, what you're saying makes sense. I'll make those changes asap.

sweeneyde commented on CrossPlatformDesktopProject/CrossPlatformDesktopProject/NPC/Npc.cs:
```diff
@@ -0,0 +1,37 @@
+namespace CrossPlatformDesktopProject.NPC
+{
+    class Npc
+    {
```
I think it's confusing that there's an `INpc` interface that `Npc` doesn't implement.

I wonder if it would be better to have an `INpc` interface implemented by a `Goriya` class, and `Goriya` objects store `IGoriyaState`s, where things like `GoriyaWalkSouth` implement `IGoriyaState`? Maybe this would give a nicer way to construct a `new Goriya()` by calling its constructor, rather than by constructing a general NPC.

royjpark replied:
So do you mean like having a separate `Goriya` class from `GoriaWalkSouth` and etc.? And depending on `IGoriyaState` it will call the required direction? Sorry I'm a bit confused.

sweeneyde replied:
I was thinking
```
interface INpc {...}

interface IGoriyaState {...}

class Goriya : INpc {
    public IGoriyaState currentState;
    ...
}

class GoriyaWalkSouth : IGoriyaState {...}

... [other Goriya states] ...
```

and then

```
interface IBatState {...}

class Bat : INpc
{
    public IBatState currentState;
    ...
}

class BatFlyingSouthOrWhatever : IBatState {...}

... [other Bat states] ...
```

I was just thinking that there's no reason a Goriya would have a Bat state, so making those interfaces only per-enemy type would be more enapsulated. I could be wrong though, so feel free to disagree.

royjpark replied:
I just made the changes, hopefully I understood you right lol. If not, just let me know and I'll fix whatever. Sorry for taking so long, had an exam this morning that I was studying for.

royjpark committed:
Deleted Item class and made necessary adjustments to the Heart and Triforce classes.

royjpark committed:
Deleted Npc class and made necessary adjustments

royjpark committed:
idk

royjpark committed:
Merged w master

royjpark committed:
MErged

royjpark commented:
Update: 
- Deleted Npc class
- Added interfaces for each enemy's state
- Created a class for each enemy that implements iNPC
- Changed Old Man NPC to be more like a World Item class because there's no movement/states.

sweeneyde commented:
I haven't tried running it but I think the changes seem good.

sweeneyde commented:
Update: It runs fine on my machine. LGTM
